### PR TITLE
custom paths in settings schema

### LINKED ISSUE #303: buildtest settings update
@vsoch i am trying to work on another revision for buildtest settings file, here is an overview of some of the changes, i wanted to propose and would like your input. I will work on the schema in the meantime but figured i shared this now since changes to YAML will require changes to json schema.

Screen Shot 2020-05-26 at 9 03 38 PM

The `printenv` section is an executor specific configuration that allows all tests using the given executor to have environment variables injected into job script. The printenv is just going to echo the environment variable as simple as `echo $FOO`

The `config` section is where the core of buildtest configuration is setup, so this where you can configure `tests` `slurm` `paths` and `vars` which are configuration sections somehow related to buildtest. For instance `tests`  can have a timeout in seconds, if test fails to complete in given time this would be marked as fail. The `paths` where we define buildtest configuration, if it's not defined buildtest will resort to default behavior what we have. Setting these paths would be nice to have full control where user wants to store files depending on the filesystem preference (paralllel file-system, scratch, home, etc...)

The `vars` section is used to define any global variables accessible by all tests, the idea is to have a location to define a shared variable and this will be sourced into script or automatically added into script. 

Each of the global configuration in `config` could be overridden if defined in `executors` section which would be associated to a particular test.

The `slurm` section is where we define slurm configuration shared by all slurm executor, at the moment, i know i will have to poll a slurm job via `sacct` to periodically check if job has finished before buildtest reports test is complete. I can remove this for now and maybe implement this when i get slurm executor working. It is on my TODO list asap.

What do you think of this setup?

### LINKED ISSUE #304: show settings schema via command line
currently all schemas are shown in `buildtest show schema` and currently we have the following options

```
$ buildtest show schema --help
usage: buildtest [options] [COMMANDS] show schema [-h] [-v VERSION] [-n {script,compiler}] [-g]

optional arguments:
  -h, --help            show this help message and exit
  -v VERSION, --version VERSION
                        choose a specific version of schema to show.
  -n {script,compiler}, --name {script,compiler}
                        show schema by name (e.g., script)
  -g, --global          show global schema
```

Since the settings schema is not versioned we could put a new option ``-s`` with long option ``--settings`` to show ``settings.schema.json``. Currently this file is located in https://github.com/buildtesters/buildtest/tree/devel/buildtest/settings while all schemas are in https://github.com/buildtesters/buildtest/tree/devel/buildtest/buildsystem/schemas 

If we move the file the second directory it would simplify the lookup but few variables would have to be changed in defaults.py

@vsoch what do you think about this feature? Should we move the files under schemas directory, and what about the option `--settings``?

### PR DESCRIPTION
**Pull request by shahzebsiddiqui**
Summary
-----------

- Adding option `buildtest show schema --settings` to show settings schema
- Add `config` section in settings schema with subsection `paths` where we can define the test directory, location where to search buildspecs which is a colon separate list that mimics $PATH but for buildspecs, and location where to clone repos via `buildtest get`.

### CODE REVIEW

**Comment by shahzebsiddiqui** [comment]
Here is my configuration file

```
(buildtest) siddiq90@DOE-7086392 buildtest % buildtest config view
editor: vi
executors:
  runlocal:
    type: local
    description: submit jobs locally
config:
  paths:
    get_repo_path: /tmp/buildtest/repos
    testdir: /tmp/buildtest/tests
    buildspec_path: /tmp/buildtest/repos
```

Custom Clone Path get_repo_path
-------------------------------------

```
(buildtest) siddiq90@DOE-7086392 buildtest % buildtest get https://github.com/buildtesters/buildtest-cori
Cloning into '/private/tmp/buildtest/repos/github.com/buildtesters/buildtest-cori'...
remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 9 (delta 0), reused 9 (delta 0), pack-reused 0
Unpacking objects: 100% (9/9), done.
```

If `get_repo_path` is not set it will default to original path as implemented before

```
(buildtest) siddiq90@DOE-7086392 buildtest % buildtest get https://github.com/buildtesters/buildtest-cori                 
Cloning into '/Users/siddiq90/.buildtest/site/github.com/buildtesters/buildtest-cori'...
remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 9 (delta 0), reused 9 (delta 0), pack-reused 0
Unpacking objects: 100% (9/9), done.
```

Customize testdir
--------------------

If testdir defined in config it will take precedence, here is a sample testdir when running `buildtest build -b tests/examples/buildspecs/environment.yml`

```
buildtest) siddiq90@DOE-7086392 buildtest % ls -l /tmp/buildtest/tests
total 8
-rwxr-xr-x  1 siddiq90  wheel   87 May 27 08:45 hello_dinosaur.sh
drwxr-xr-x  6 siddiq90  wheel  192 May 27 08:45 run
```

If --testdir passed on command line, this overrides configuration.
```
buildtest) siddiq90@DOE-7086392 buildtest % buildtest build -b tests/examples/buildspecs/environment.yml --testdir /tmp/random

            Discovered Buildspecs              

/Users/siddiq90/Documents/buildtest/tests/examples/buildspecs/environment.yml



Buildspec Name                 SubTest                        Status                         Buildspec Path                
________________________________________________________________________________________________________________________
environment                    hello_dinosaur                 PASS                           /Users/siddiq90/Documents/buildtest/tests/examples/buildspecs/environment.yml


============================================================
                        Test summary                        
============================================================
Executed 1 tests
Passed Tests: 1/1 Percentage: 100.000%
Failed Tests: 0/1 Percentage: 0.000%
(buildtest) siddiq90@DOE-7086392 buildtest % ls -l /tmp/random
total 8
-rwxr-xr-x  1 siddiq90  wheel   87 May 27 08:48 hello_dinosaur.sh
drwxr-xr-x  4 siddiq90  wheel  128 May 27 08:48 run
```

Buildspec Paths
-----------------

This is probably the most useful feature and i think having user be able to customize search path will help autocomplete full path by passing relative path. So with original configuration as mentioned above we can reference a buildspec now using the search path

```
(buildtest) siddiq90@DOE-7086392 buildtest % buildtest build -b github.com/buildtesters/buildtest-cori/system/mountpoint.yml

            Discovered Buildspecs              

/private/tmp/buildtest/repos/github.com/buildtesters/buildtest-cori/system/mountpoint.yml



Buildspec Name                 SubTest                        Status                         Buildspec Path                
________________________________________________________________________________________________________________________
mountpoint                     mount_check                    FAIL                           /private/tmp/buildtest/repos/github.com/buildtesters/buildtest-cori/system/mountpoint.yml


============================================================
                        Test summary                        
============================================================
Executed 1 tests
Passed Tests: 0/1 Percentage: 0.000%
Failed Tests: 1/1 Percentage: 100.000%
```

We can add extra search paths so that user don't have to specify the full path relative from github.com ... such as one below

```
(buildtest) siddiq90@DOE-7086392 buildtest % cat ~/.buildtest/settings.yml| grep buildspec
    buildspec_path: /tmp/buildtest/repos/github.com/buildtesters/buildtest-cori:/private/tmp/buildtest/repos/github.com/buildtesters/buildtest-stampede2
```

```
(buildtest) siddiq90@DOE-7086392 buildtest % buildtest build -b system/mountpoint.yml

            Discovered Buildspecs              

/private/tmp/buildtest/repos/github.com/buildtesters/buildtest-cori/system/mountpoint.yml



Buildspec Name                 SubTest                        Status                         Buildspec Path                
________________________________________________________________________________________________________________________
mountpoint                     mount_check                    FAIL                           /private/tmp/buildtest/repos/github.com/buildtesters/buildtest-cori/system/mountpoint.yml


============================================================
                        Test summary                        
============================================================
Executed 1 tests
Passed Tests: 0/1 Percentage: 0.000%
Failed Tests: 1/1 Percentage: 100.000%
```

Basically we get the first entry in search path that is a hit and process the buildspec which is how $PATH behaves. If there are two files with same name then whichever search tree was defined first will be honored. Since `buildspec_path` is not required if its not defined it will honor the default path `$HOME/.buildtest/site`

File: buildtest/menu/build.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 logger = logging.getLogger(__name__)
 
 
-def discover_buildspecs(buildspec):
+def discover_buildspecs(buildspec, search_path=[]):
```
Comment (vsoch) [inline buildtest/menu/build.py]: You should generally not set a list as an input argument to a function, it's mutable. Instead this should be set to None, and then `search_path = search_path or []`

File: buildtest/menu/build.py [hunk 2]
```diff
@@ -38,15 +38,17 @@ def discover_buildspecs(buildspec):
...
-    if not buildspec:
-        buildspec = BUILDSPEC_DEFAULT_PATH
+    # add default path to search path
+    search_path.append(BUILDSPEC_DEFAULT_PATH)
 
     # First try, the path is an absolute path to file or folder
     # Second try, the path can be relative to the BUILDSPEC_DEFAULT_PATH
-    elif not os.path.exists(buildspec):
-        buildspec = os.path.join(BUILDSPEC_DEFAULT_PATH, buildspec)
+    if not os.path.exists(buildspec):
```
Comment (vsoch) [inline buildtest/menu/build.py:47]: Why did you remove the check here if it's set to None? This is again going to trigger an error if this is the case.

File: buildtest/menu/build.py [hunk 3]
```diff
@@ -64,9 +66,10 @@ def discover_buildspecs(buildspec):
         logger.debug(f"BuildSpec: {buildspec} is a file")
     else:
         msg = (
-            "Please provide an absolute or relative path to a directory file from your present working directory or %s"
-            % BUILDSPEC_DEFAULT_PATH
+            f"Unable to find any buildspecs in search paths: {search_path} \n"
+            + "Please provide an absolute or relative path to a directory or file or file relative to current directory."
```
Comment (vsoch) [inline buildtest/menu/build.py]: or file or file

File: buildtest/menu/build.py [hunk 4]
```diff
@@ -148,13 +151,32 @@ def func_build_subcmd(args):
...
 
+    config_paths_testdir = config_opts.get("config", {}).get("paths", {}).get("testdir")
+    config_paths_buildspec_path = (
+        config_opts.get("config", {}).get("paths", {}).get("buildspec_path")
+    )
+    buildspec_paths = []
+    if config_paths_buildspec_path:
+        # return a unique list of directory paths where buildtest should search for buildspecs
+        # this is equivalent to $PATH for buildspec
+        buildspec_paths = list(set(config_paths_buildspec_path.split(":")))
```
Comment (vsoch) [inline buildtest/menu/build.py]: No need to strip for empty spaces (e.g., if there is a trailing : you are going to get an empty string as a path which will have unintended consequences...

**Review by vsoch — CHANGES_REQUESTED** [review]

File: buildtest/menu/build.py (hunk 2 above)
Comment (shahzebsiddiqui) [reply buildtest/menu/build.py:47]: Well i don't anticipate user will pass `buildtest build` we shouldnt even allow this. They should pass with `-b` option. That's why i removed the check

File: buildtest/menu/build.py (hunk 4 above)
Comment (shahzebsiddiqui) [reply buildtest/menu/build.py]: I see the trailing `:` is an issue i can put a check and remove it if found at end of char, we okay with splitting by colon i assume since this check would be done by default.

Comment (shahzebsiddiqui) [reply buildtest/menu/build.py]: i tested this locally and got this resolved, see latest commit

**Comment by vsoch** [comment]
@shahzebsiddiqui I won't have time in the next month or two to continue reviewing, it's very arduous. Please find someone else to review in the meantime. Thanks!

**Comment by shahzebsiddiqui** [comment]
Thank you for your time and contribution to buildtest, your feedback has been very helpful and insightful over the past 3 months. Sure, i will figure something out in the interim, i can totally understand your situation i don't have the bandwidth to contribute to projects like you do. I wish you the best, let's stay in touch :-)
